### ISSUE TITLE
CPU caches not reported on illumos

### ISSUE THREAD

**Issue by rzezeski** [open-mpi/hwloc#144]
The `lstopo` output does not report CPU cache information on illumos-based x86 operating systems (SmartOS in my particular case). While studying the debug output and with a little help form DTrace I was able to determine that hwloc is able to discover the cache information without problem, it just doesn't print it. I found that if I modified the code to set `fulldiscovery=1` in `look_procs()`, before the call to `summarize()`, that the cache information is displayed. Knowing this obviously isn't a proper fix I tried to dig a bit more and believe I narrowed the issue to a condition in `hwloc_x86_discover()` [1]. If `fulldiscovery` is not taken then the cache information is not output.

1: https://github.com/open-mpi/hwloc/blob/master/hwloc/topology-x86.c#L1222

```C
    if (topology->nb_levels == 2 && topology->level_nbobjects[1] == data->nbprocs) {
      /* only PUs were discovered, as much as we would, complete the topology with everything else */
      alreadypus = 1;
      goto fulldiscovery;
    }
```

With no familiarity of the hwloc code I am left at the mercy of the code comment, which is hard to parse and doesn't leave me with much to go on. I think it's saying that if we've only discovered processors that we need to perform fulldiscovery. That makes sense, but when I run `lstopo` it reports only 1 level at this point.

```
topology->nb_levels = 1
```

It would be great if someone familiar with this code could point me in the right direction. I'm very familiar with illumos and can answer any questions there. But I don't have any depth in hwloc and would prefer not to spend hours learning it when someone might be able to spot the issue in a few minutes.

Here is the current output:

```
Machine (32GB total) + NUMANode L#0 (P#0 32GB) + Package L#0
  Core L#0
    PU L#0 (P#0)
    PU L#1 (P#1)
  Core L#1
    PU L#2 (P#2)
    PU L#3 (P#3)
  Core L#2
    PU L#4 (P#4)
    PU L#5 (P#5)
  Core L#3
    PU L#6 (P#6)
    PU L#7 (P#7)
  Core L#4
    PU L#8 (P#8)
    PU L#9 (P#9)
  Core L#5
    PU L#10 (P#10)
    PU L#11 (P#11)
```

Here is the output with the `fulldiscovery=1` hack:

```
Machine (32GB total) + NUMANode L#0 (P#0 32GB) + Package L#0 + L3 L#0 (15MB)
  L2 L#0 (256KB) + L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0
    PU L#0 (P#0)
    PU L#1 (P#1)
  L2 L#1 (256KB) + L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1
    PU L#2 (P#2)
    PU L#3 (P#3)
  L2 L#2 (256KB) + L1d L#2 (32KB) + L1i L#2 (32KB) + Core L#2
    PU L#4 (P#4)
    PU L#5 (P#5)
  L2 L#3 (256KB) + L1d L#3 (32KB) + L1i L#3 (32KB) + Core L#3
    PU L#6 (P#6)
    PU L#7 (P#7)
  L2 L#4 (256KB) + L1d L#4 (32KB) + L1i L#4 (32KB) + Core L#4
    PU L#8 (P#8)
    PU L#9 (P#9)
  L2 L#5 (256KB) + L1d L#5 (32KB) + L1i L#5 (32KB) + Core L#5
    PU L#10 (P#10)
    PU L#11 (P#11)
```

**Comment by bgoglin** [comment]
I am not familiar with Illumos. If that's a Solaris-like system with x86 processors, the problem is likely that the x86 backend (topology-x86.c) should complete the partial discovery made by Solaris backend earlier (topology-solaris.c found NUMA/package/core/PU information but no cache information). That partial discovery isn't enabled because these backends could conflict for objects they both discover (packages and cores, if the BIOS and/or kernel is buggy).

Try passing HWLOC_COMPONENTS=x86 in the environment enables the x86 backend first, that should report a topology with packages/cores/caches but no NUMA (except on recent AMD processors).

I'll see if I can enable partial x86 discovery in your particular case (when we only want to add caches).

**Comment by bgoglin** [comment]
By the way, other Solaris-like systems have the same problem. We don't ever detect caches there anyway.

**Comment by rzezeski** [comment]
> I am not familiar with Illumos. If that's a Solaris-like system

Yes, illumos is a community fork of OpenSolaris started circa 2010. Oracle Solaris and illumos have diverged since then but a lot of the stuff hwloc relies on should be the same: e.g. kstats, lgrps, process groups, psrsets, etc.

> with x86 processors, the problem is likely that the x86 backend (topology-x86.c) should complete the partial discovery made by Solaris backend earlier (topology-solaris.c found NUMA/package/core/PU information but no cache information). That partial discovery isn't enabled because these backends could conflict for objects they both discover (packages and cores, if the BIOS and/or kernel is buggy).

> Try passing HWLOC_COMPONENTS=x86 in the environment enables the x86 backend first, that should report a topology with packages/cores/caches but no NUMA (except on recent AMD processors).

This information helps a lot. Thanks. I hope to have time to dig back into this later in the evening.

**Comment by bgoglin** [comment]
Please just test this tarball *without* setting any environment variable as said above:
https://ci.inria.fr/hwloc/job/zcustombranch-0-tarball/lastSuccessfulBuild/artifact/hwloc-v1.11-20160114.2318.gitcbaf37c.tar.gz
It adds the ability to complete the Solaris-backend discovery with x86-backend caches. Quick testing on OpenIndiana says it works fine.

**Comment by rzezeski** [comment]
> It adds the ability to complete the Solaris-backend discovery with x86-backend caches. Quick testing on OpenIndiana says it works fine.

I tried your patch on two different machines. Everything looks good.

```
Machine (32GB total) + NUMANode L#0 (P#0 32GB) + Package L#0 + L3 L#0 (15MB)
  L2 L#0 (256KB) + L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0
    PU L#0 (P#0)
    PU L#1 (P#1)
  L2 L#1 (256KB) + L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1
    PU L#2 (P#2)
    PU L#3 (P#3)
  L2 L#2 (256KB) + L1d L#2 (32KB) + L1i L#2 (32KB) + Core L#2
    PU L#4 (P#4)
    PU L#5 (P#5)
  L2 L#3 (256KB) + L1d L#3 (32KB) + L1i L#3 (32KB) + Core L#3
    PU L#6 (P#6)
    PU L#7 (P#7)
  L2 L#4 (256KB) + L1d L#4 (32KB) + L1i L#4 (32KB) + Core L#4
    PU L#8 (P#8)
    PU L#9 (P#9)
  L2 L#5 (256KB) + L1d L#5 (32KB) + L1i L#5 (32KB) + Core L#5
    PU L#10 (P#10)
    PU L#11 (P#11)
```

```
# /opt/local/hwloc-bgoglin/bin/lstopo
Machine (512GB total)
  Package L#0
    NUMANode L#0 (P#1 128GB) + L3 L#0 (20MB)
      L2 L#0 (256KB) + L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0
        PU L#0 (P#0)
        PU L#1 (P#32)
      L2 L#1 (256KB) + L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1
        PU L#2 (P#2)
        PU L#3 (P#34)
      L2 L#2 (256KB) + L1d L#2 (32KB) + L1i L#2 (32KB) + Core L#2
        PU L#4 (P#4)
        PU L#5 (P#36)
      L2 L#3 (256KB) + L1d L#3 (32KB) + L1i L#3 (32KB) + Core L#3
        PU L#6 (P#6)
        PU L#7 (P#38)
      L2 L#4 (256KB) + L1d L#4 (32KB) + L1i L#4 (32KB) + Core L#4
        PU L#8 (P#8)
        PU L#9 (P#40)
      L2 L#5 (256KB) + L1d L#5 (32KB) + L1i L#5 (32KB) + Core L#5
        PU L#10 (P#10)
        PU L#11 (P#42)
      L2 L#6 (256KB) + L1d L#6 (32KB) + L1i L#6 (32KB) + Core L#6
        PU L#12 (P#12)
        PU L#13 (P#44)
      L2 L#7 (256KB) + L1d L#7 (32KB) + L1i L#7 (32KB) + Core L#7
        PU L#14 (P#14)
        PU L#15 (P#46)
    NUMANode L#1 (P#2 128GB) + L3 L#1 (20MB)
      L2 L#8 (256KB) + L1d L#8 (32KB) + L1i L#8 (32KB) + Core L#8
        PU L#16 (P#16)
        PU L#17 (P#48)
      L2 L#9 (256KB) + L1d L#9 (32KB) + L1i L#9 (32KB) + Core L#9
        PU L#18 (P#18)
        PU L#19 (P#50)
      L2 L#10 (256KB) + L1d L#10 (32KB) + L1i L#10 (32KB) + Core L#10
        PU L#20 (P#20)
        PU L#21 (P#52)
      L2 L#11 (256KB) + L1d L#11 (32KB) + L1i L#11 (32KB) + Core L#11
        PU L#22 (P#22)
        PU L#23 (P#54)
      L2 L#12 (256KB) + L1d L#12 (32KB) + L1i L#12 (32KB) + Core L#12
        PU L#24 (P#24)
        PU L#25 (P#56)
      L2 L#13 (256KB) + L1d L#13 (32KB) + L1i L#13 (32KB) + Core L#13
        PU L#26 (P#26)
        PU L#27 (P#58)
      L2 L#14 (256KB) + L1d L#14 (32KB) + L1i L#14 (32KB) + Core L#14
        PU L#28 (P#28)
        PU L#29 (P#60)
      L2 L#15 (256KB) + L1d L#15 (32KB) + L1i L#15 (32KB) + Core L#15
        PU L#30 (P#30)
        PU L#31 (P#62)
  Package L#1
    NUMANode L#2 (P#5 128GB) + L3 L#2 (20MB)
      L2 L#16 (256KB) + L1d L#16 (32KB) + L1i L#16 (32KB) + Core L#16
        PU L#32 (P#1)
        PU L#33 (P#33)
      L2 L#17 (256KB) + L1d L#17 (32KB) + L1i L#17 (32KB) + Core L#17
        PU L#34 (P#3)
        PU L#35 (P#35)
      L2 L#18 (256KB) + L1d L#18 (32KB) + L1i L#18 (32KB) + Core L#18
        PU L#36 (P#5)
        PU L#37 (P#37)
      L2 L#19 (256KB) + L1d L#19 (32KB) + L1i L#19 (32KB) + Core L#19
        PU L#38 (P#7)
        PU L#39 (P#39)
      L2 L#20 (256KB) + L1d L#20 (32KB) + L1i L#20 (32KB) + Core L#20
        PU L#40 (P#9)
        PU L#41 (P#41)
      L2 L#21 (256KB) + L1d L#21 (32KB) + L1i L#21 (32KB) + Core L#21
        PU L#42 (P#11)
        PU L#43 (P#43)
      L2 L#22 (256KB) + L1d L#22 (32KB) + L1i L#22 (32KB) + Core L#22
        PU L#44 (P#13)
        PU L#45 (P#45)
      L2 L#23 (256KB) + L1d L#23 (32KB) + L1i L#23 (32KB) + Core L#23
        PU L#46 (P#15)
        PU L#47 (P#47)
    NUMANode L#3 (P#6 128GB) + L3 L#3 (20MB)
      L2 L#24 (256KB) + L1d L#24 (32KB) + L1i L#24 (32KB) + Core L#24
        PU L#48 (P#17)
        PU L#49 (P#49)
      L2 L#25 (256KB) + L1d L#25 (32KB) + L1i L#25 (32KB) + Core L#25
        PU L#50 (P#19)
        PU L#51 (P#51)
      L2 L#26 (256KB) + L1d L#26 (32KB) + L1i L#26 (32KB) + Core L#26
        PU L#52 (P#21)
        PU L#53 (P#53)
      L2 L#27 (256KB) + L1d L#27 (32KB) + L1i L#27 (32KB) + Core L#27
        PU L#54 (P#23)
        PU L#55 (P#55)
      L2 L#28 (256KB) + L1d L#28 (32KB) + L1i L#28 (32KB) + Core L#28
        PU L#56 (P#25)
        PU L#57 (P#57)
      L2 L#29 (256KB) + L1d L#29 (32KB) + L1i L#29 (32KB) + Core L#29
        PU L#58 (P#27)
        PU L#59 (P#59)
      L2 L#30 (256KB) + L1d L#30 (32KB) + L1i L#30 (32KB) + Core L#30
        PU L#60 (P#29)
        PU L#61 (P#61)
      L2 L#31 (256KB) + L1d L#31 (32KB) + L1i L#31 (32KB) + Core L#31
        PU L#62 (P#31)
        PU L#63 (P#63)
  HostBridge L#0
    PCIBridge
      2 x { PCI 8086:154d }
    PCIBridge
      2 x { PCI 8086:10fb }
    PCI 8086:8d62
    PCIBridge
      2 x { PCI 8086:1521 }
    PCIBridge
      PCIBridge
        PCIBridge
          PCIBridge
            PCI 102b:0534
    PCI 8086:8d02
```

**Comment by bgoglin** [comment]
This will be fixed in 1.11.3 and 2.0
